src/storage/src/upsert.rs

```diff
@@ -670,259 +850,341 @@ where
         &output_handle,
     );
 
-    let mut previous = builder.new_input_for(
-        &previous.inner,
+    let mut persist_input = builder.new_disconnected_input(
+        &persist_input.inner,
         Exchange::new(|((key, _), _, _)| UpsertKey::hashed(key)),
-        &output_handle,
     );
 
     let upsert_shared_metrics = Arc::clone(&upsert_metrics.shared);
+
     let shutdown_button = builder.build(move |caps| async move {
-        let [mut output_cap, mut snapshot_cap, health_cap]: [_; 3] = caps.try_into().unwrap();
+        let [mut output_cap, snapshot_cap, health_cap]: [_; 3] = caps.try_into().unwrap();
+        let mut snapshot_cap = CapabilitySet::from_elem(snapshot_cap);
 
         // The order key of the `UpsertState` is `Option<FromTime>`, which implements `Default`
         // (as required for `consolidate_snapshot_chunk`), with slightly more efficient serialization
         // than a default `Partitioned`.
+
         let mut state = UpsertState::<_, Option<FromTime>>::new(
-            state().await,
-            upsert_shared_metrics,
+            state_fn().await,
+            Arc::clone(&upsert_shared_metrics),
             &upsert_metrics,
-            source_config.source_statistics,
+            source_config.source_statistics.clone(),
             upsert_config.shrink_upsert_unused_buffers_by_ratio,
         );
-        let mut events = vec![];
-        let mut snapshot_upper = Antichain::from_elem(Timestamp::minimum());
 
+        // True while we're still reading the initial "snapshot" (a whole bunch
+        // of updates, all at the same initial timestamp) from our persist
+        // input.
+        let mut snapshotting_persist = true;
```

Comment:
I didn't understand what it means to snapshot persist. Is this flag tracking whether we are rehydrating or not?

Reply:
> Is this flag tracking whether we are rehydrating or not?

It is, sometimes we call that snapshotting, or "reading the snapshot", hence the name. But I can rename this to `rehydrating`.